### ISSUE TITLE
Missing clone feature on WidgetMessage: tm-new-tiddler (or documentation missing)

### ISSUE THREAD

**Issue by AnthonyMuscio** [Jermolene/TiddlyWiki5#4411]
We can currently clone (Copy the whole tiddler tags, text and fields to a new one) a tiddler with the clone button or the following;
`<$button message="tm-new-tiddler" param="TestTiddler"> Clone TestTiddler</$button>`
This process will increment the source tiddler title to create the new one.

This is a useful default but only represents a subset of the desired clone function. 

I would like to be able to provide an alternative title which if it exists will also be incremented. 

There is reference to the _paramObject as an Optional hashmap of additional tiddler fields_
https://tiddlywiki.com/?user=test#WidgetMessage%3A%20tm-new-tiddler
and _If a hashmap was used and a title field was specified, use that title_

However there is not documentation on how to use a hashmap anywhere in tiddlywiki, if you think it simply means provide the parameter title="new title" as in the subsequent example that will not work.

Unless this is resolved/Documented/worked around, I will have to build a method to copy select fields and values from the source tiddler to the destination.

Regards
Tony

**Comment by pmario** [comment]
The button-widget doesn't create a paramObject see: https://tiddlywiki.com/?user=test#ButtonWidget It only knows the parameters that are listed in the docs. 

That's why the action-widgets have been designed, to be more flexible. The problem is, that action-widgets and the button-widget functions are not 100% consistent. 

If you use the following test-command it almost does what you describe. BUT if the `title` is provided it will be used. There is no additional check if the tiddler exists. 

```
<$button>
<$action-sendmessage $message="tm-new-tiddler"  $param=asdf title="hugo" tags="OneTag [[Another Tag]]" text=<<now "Today is DDth, MMM YYYY">>/>
New Tiddler
</$button>
```

We can _not_ change this behaviour, since it may break existing stuff. 

What we could do, would be a new parameter, eg: "check for existing tiddler and create a new name if it exists" parameter. ... but shorter ;)

**Comment by pmario** [comment]
The documentation says it all. see: https://tiddlywiki.com/prerelease/#WidgetMessage%3A%20tm-new-tiddler  ... section: "The title for the draft tiddler is chosen according to these rules:"

**Comment by AnthonyMuscio** [comment]
Mario,

I think it would be a simple matter of updating the documentation. Cloning tiddlers to a new title is a desirable feature, it provides a more expensive tiddler templating than simple tiddlywiki templates.  Thanks to you comments, which I admit I cant follow all the details I found this to work; To clone TemplateTiddler as DestinationTiddler.

```
<$button>
<$action-sendmessage $message="tm-new-tiddler"  $param="TemplateTiddler" title="DestinationTiddler"/>
Clone Tiddler
</$button>
```
With the caveat that it will overwrite DestinationTiddler would be fine.

It can simply be wrapped in a test to only show if the DestinationTiddler does not exist, or search a range until an increment there of is not found.

Thanks
Tony

**Comment by pmario** [comment]
You should: 

```
\define xxActions()
<$action-sendmessage $message="tm-new-tiddler"  $param="TemplateTiddler" title="DestinationTiddler"/>
<!-- more actions go here -->
\end

<$button actions=xxActions>
Clone Tiddler
</$button>
```

With this contstructioin the action variables are updated on button-click. 

If you do it your way, the variables are only updated on button-render, which can cause problems. ... This was the main reason, why the actions-parameter has been introduced for buttons.

**Comment by pmario** [comment]
fixed typo in the example above.

**Comment by pmario** [comment]
The point is, that I do want to have a "real" clone function, where I can decide if I need "unique" names AND if the tiddler should be opened automatically. ... 

So I think the issue is still valid, even if you can do a workaround. ... The usecase you described, imo shouldn't need a workaround. 

At the moment I'm also struggling with a "tiddler clone" problem, where I do want to clone an existing tiddler as `$:/temp/story` to be able to have "edit / undo" functionality.

**Comment by pmario** [comment]
There is the https://tiddlywiki.com/prerelease/#ActionCreateTiddlerWidget which could be extended to get `$template, $unique and $open` parameters. 

 - `$template` ... name of the template tiddler
    - if NO $basetitle is there, it uses the template name as $basetitle
 - `$unique` 
    - `yes` (default) see `$basetitle` 
    - `no` $basetitle is used and fixed. (parameter is "no" or doesn't exist)
 - `$open` 
    - `no` is default -> don't open (parameter is "no" or doesn't exist)
    - `yes` open in view mode
    - `edit` open in edit mode

It should be simple to implement this. 

@Jermolene  what do you think? I would create a PR

**Comment by pmario** [comment]
Just found out, that the $open parameter is not good, since it creates too much duplicated code. ActionSendMessage widget can be used instead.

**Comment by AnthonyMuscio** [comment]
Mario. If we use the actions= method we can simply add "to=tiddlername" to a button or an action navigate or tm-navigate the problem is that tm-new-tiddler assumes too much and also navigates to edit the tiddler.

I notice that if I add additional field value pairs the clone function not longer works. perhaps we should copy tm-new-tiddler as tm-clone-tiddler and modify it to just do it, and demand a separate navigation step. This will simplify the parameters eg
`<$action-sendmessage $message="tm-clone-tiddler"  $source="TemplateTiddler" $destination="DestinationTiddler"/>`

This will support the bulk cloning of tiddlers if needed. Perhaps it can also skip the creation of a draft tiddler.

Regards
Tony

**Comment by AnthonyMuscio** [comment]
I believe this has being addressed in the latest versions

**Comment by kookma** [comment]
> There is the https://tiddlywiki.com/prerelease/#ActionCreateTiddlerWidget which could be extended to get `$template, $unique and $open` parameters.
> 
> * `$template` ... name of the template tiddler
>   
>   * if NO $basetitle is there, it uses the template name as $basetitle
> * `$unique`
>   
>   * `yes` (default) see `$basetitle`
>   * `no` $basetitle is used and fixed. (parameter is "no" or doesn't exist)
>     
>     * $basetitle must be provided
> * `$open`
>   
>   * `no` is default -> don't open (parameter is "no" or doesn't exist)
>   * `yes` open in view mode
>   * `edit` open in edit mode
> 
> It should be simple to implement this.
> 
> @Jermolene what do you think? I would create a PR

@pmario 
This is very useful solution and simplify and unify most required button behavior in one widget!
Why not submit a PR?

**Comment by pmario** [comment]
@kookma most of those ideas have been implemented in v5.1.22. See. $template and $overwrite.

There was an issue with the $savetitle parameter, which doesn't work as expected, if action-createtiddler was followed by other actions, that needed to use the parameter. The fix comes with v5.2.0

The following code works with the prerelease

It creates a new tiddler based on a template and opens it. 
see: https://tiddlywiki.com/prerelease/#ActionCreateTiddlerWidget for the description of the new variables.
The important aspect here is, that the $action-sendmessage needs to be in th $action-createtiddler body to use the new "createTiddler-title varioble

```
\define testCreate()
<$action-createtiddler $basetitle="base" $template="ActionCreateTiddlerWidget Template" >
    <$action-sendmessage $message="tm-edit-tiddler" $param=<<createTiddler-title>>/>
</action-createtiddler>
\end

<$button actions=<<testCreate>> >
Create Tiddler
</$button>
```

**Comment by linonetwo** [comment]
When `action-createtiddler` called on a draft tiddler, it will resulted in a tiddler with `Draft of`
<img width="1279" alt="截屏2022-01-18 下午11 03 15" src="https://user-images.githubusercontent.com/3746270/149962579-cab0b53c-dfe4-49ce-8f76-4cc58f776253.png">
<img width="1009" alt="截屏2022-01-18 下午11 02 59" src="https://user-images.githubusercontent.com/3746270/149962594-096cbc0a-bbcb-4110-9cf5-c6974a4790f1.png">

```js
<$action-createtiddler $basetitle=<<target>> $overwrite=yes $template=<<currentTiddler>> >
          
</$action-createtiddler>
```

So I think `$savetitle` shouldn't be deprecated, it can still work in this scenario.

**Comment by linonetwo** [comment]
I know why, create-tiddler with template will overwride all the fields, so `draft.of` and `draft.title` will be overwrite too.

A fix can be:

```
          <$action-setfield $tiddler=<<target>> $field="draft.of" $value=<<draftOf>> />
          <$action-setfield $tiddler=<<target>> $field="draft.title" $value=<<targetTitle>> />
```

**Comment by Jermolene** [comment]
Thank you @linonetwo 

> So I think $savetitle shouldn't be deprecated, it can still work in this scenario. So I can make sure newly created tiddler won't have Draft of.

I don't quite understand. How does the $savetitle attribute help in the case of cloning a draft tiddler?

**Comment by linonetwo** [comment]
I was creating a button that will call \`$action-createtiddler\` in the draft mode: [https://talk.tiddlywiki.org/t/apply-template-to-your-new-tiddler-a-template-list-plugin/2121](https://talk.tiddlywiki.org/t/apply-template-to-your-new-tiddler-a-template-list-plugin/2121)

So I'm not "cloning a draft tiddler", I'm cloning a template tiddler to overwrite the current draft tiddler.

And I found that the the newly created tiddler will have wrong title, so I thought `$savetitle` will create the correct title.

But later I found I can work around this by setting `draft.of` and `draft.title` ! So I don't need `$savetitle` anymore!
